### DISCUSSION TITLE
Model subtypes

### DISCUSSION THREAD

**Discussion by exenestecnico** [adonisjs/core#1264] [Help] [answered]
How would you go about model subtypes? I have Clients that can be of two types, with shared fields and type specific fields. Already made migrations for the 3 tables.

![uml](https://user-images.githubusercontent.com/40433929/87862935-7314c200-c923-11ea-9761-a8c12f11dbf6.png)

I went thru all the forum posts in the Database section. The only [related](https://forum.adonisjs.com/t/split-product-table-into-several-tables/2716) one that I could find didn't get into managing Model subtypes in Adonis.

What should I do?:
- make 3 models with relations
- use Query Builder
- write raw queries
- make database triggers that write the data into the separate tables

**Comment by thetutlage** [comment, marked as answer]
There is no concept of sub types in Lucid. Infact, I have not seen them in majority of streamline ORMs. What you are after is relationships.

- Yes, there will be 3 models
- `Residential Client → hasOne → Client`
- `Stalal Client → hasOne → Client`

With relationships in place, The ORM output object will be as follows:

```js
{
  clientId: '',
  entityId: '',
  client: {
     type: '',
     contract
  }
}
```

You are looking to fetch a merged flat list of attributes. The ORM is all about relationships and trees.

**Reply by nelsonjd** [reply]
Thank you this is so helpful.  
However, wouldn't this be `Residential Client -> belongsTo -> Client`
`Stalal Client -> belongsTo -> Client`?
Because the foreign_key is always on the table whose model does the belongsTo.

**Comment by nelsonjd** [comment]
How would you handle a many-to-many relationship with these "clients"?  Would you do a many-to-many with `Client` and then have nullable has-one relationships between `Client -> hasOne -> Residential Client` and `Client -> hasOne -> Statal Client`? Or would you have two many-to-many relationships, one to `Residential Clients` and one to `Statal Clients`?
